src/Umbraco.Core/Cache/DefaultRepositoryCachePolicy.cs

```diff
@@ -51,6 +51,34 @@ protected virtual void InsertEntity(string cacheKey, TEntity entity)
             Cache.InsertCacheItem(cacheKey, () => entity, TimeSpan.FromMinutes(5), true);
         }
 
+        private void InsertEntityById(TEntity entity)
+        {
+            // just to be safe, we cannot cache an item without an identity
+            if (entity.HasIdentity == false)
+            {
+                return;
+            }
+
+            // insert the item by id (primary lookup)
+            InsertEntity(GetEntityCacheKey(entity.Id), entity);
+
+            // make sure the item is not cached by key (secondary lookup) to avoid old version lingering in the cache (see #2997)
+            Cache.ClearCacheItem(GetEntityCacheKey(entity.Key));
```

Comment:
I'm probaby being dumb, but I don't see this code in the old code, why is this now required?

Reply:
@nul800sebastiaan you're not being dumb :) I have basically reverted the first code changes and redone the PR. That plus the GH trouble today for sure isn't helping.

The core of the issue is that the cache by key is not updated when the cache by id is updated (e.g. on create or save). The first time around I forcefully updated the cache by key alongside the cache by id. But as @Shazwazza pointed out, that explicitly doubled the memory allocated by the cache. 

This time around I'm forcefully flushing the cache by key whenever the cache by id is updated. That forces a reload of the entity when it's subsequently requested by key, thus solving the issue.